### PR TITLE
Add the in-game Masteries system: player requests, staff review, and timed training

### PR DESCRIPTION
**Pull request by zunath**
## Summary

Moves the DM/QM-managed Masteries roleplay system fully in-game, replacing the Discord-ticket + Google-spreadsheet workflow.

- **Data + rules**: `Mastery` catalog entity (seeded with all 99 entries from the spreadsheet/wiki, staff-editable in-game), `PlayerMasteryProfile` (earned tiers, training queue, quick slots, retrain credits, append-only audit log), `MasteryRequest` (Pending→InReview→Approved/Denied/Cancelled with a player↔staff comment thread), and a pure-logic `MasteryRules` engine covering the full rule set: 14/21/28/152-day duration brackets, Quick Slots (7d, T5 −21d), 17-level cap, queue cap of 3, single-Tier-5 and single-Rare limits, off-limit blocking, retrain credits (14d/7d), and eligibility (character age ≥ 14 days, rank 50 in the linked skill).
- **Player window** (`GuiWindowType.Masteries`, opened from a new Character Sheet actions-rail button): My Masteries with live training progress, searchable catalog with per-row requests (off-limit entries visible but disabled), unlisted/custom mastery requests, pre-submit eligibility warnings, and per-request comment threads with cancel.
- **Staff tooling**: `/masteryreview` (DM/Admin) — paged review queue with the same rules-check panel players see, quick-slot/instant-grant toggles with live duration preview, approve (override reason required when checks fail), deny (feedback required), plus a catalog-management tab; and a new Masteries tab on DM Player Examine for direct grants, revokes, time reductions, queue reordering, quick-slot awards, and the audit log.
- **Plumbing**: Discord embed per submission via new `SWLOR_MASTERY_STAFF_WEBHOOK_URL` env var (thread-per-request, skipped silently when unset), and lazy real-time training completion evaluated on login/window-open with persisted completion notices so offline completions are never lost.

An adversarial review pass over the finished branch surfaced 9 findings (dead retrain-credit path, instant grants queueing behind in-flight training, T5/Rare caps blind to queued training, DM examine tab blanking after modals, per-login DB writes, droppable toasts, and three minor edge cases) — all fixed with dedicated test coverage.

Deliberately deferred: one-off importer for the existing spreadsheet data (separate follow-up).

## Test plan

- 89 unit tests over `MasteryRules`/catalog seed (duration brackets, caps, retrain credits, quick-slot refunds, queue evaluation, seed idempotency) — all green.
- Full suite: 922/923; the single failure (`ForceDarkRavagerTests.ForceDarkRavagerStatusEffects_MatchCombatBible`) is pre-existing on `feature/combat-upgrade` and unrelated.
- Zero file overlap with `feature/combat-upgrade` commits made since branching.
- Remaining manual verification: NUI layout/spacing on a live client for the three new windows, and end-to-end flow against a live Redis-backed server.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added the Masteries system with a player **Masteries** window for browsing/searching the catalog, submitting requests (including custom entries), and tracking training/queue progress, tier gains, Quick Slots, and retrain credits.
  * Added staff **Mastery Review** tools to review requests, approve/deny, comment, and manage catalog entries (including Discord notifications).
  * Added Masteries login notifications and an added **Masteries** tab in the staff player-examine view.
* **Bug Fixes**
  * Improved mastery catalog seeding to prevent duplicates and preserve staff-edited entries.
  * Strengthened training-queue processing, validation, and refund/reorder behaviors.

### CODE REVIEW

**Comment by zunath** [comment]
@coderabbitai - full review

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs [hunk 1]
```diff
@@ -378,5 +594,397 @@ public Action OnClickSaveChanges() => () =>
...
+            // Earned-only rows first (alphabetical), then the active training entry, then
+            // queued entries - matching the mockup's row ordering.
+            _masteryRows.Sort((a, b) =>
+            {
+                var groupA = !a.QueueIndex.HasValue ? 0 : a.QueueIndex.Value == 0 ? 1 : 2;
+                var groupB = !b.QueueIndex.HasValue ? 0 : b.QueueIndex.Value == 0 ? 1 : 2;
+                return groupA != groupB
+                    ? groupA.CompareTo(groupB)
+                    : string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase);
+            });
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs:660]: Fixed in 463b8db53: the sort comparator now orders queued (group-2) rows by QueueIndex instead of alphabetically, so display order matches the actual queue and Move Up/Down acts on the expected neighbor.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs [hunk 2]
```diff
@@ -378,5 +594,397 @@ public Action OnClickSaveChanges() => () =>
...
+        {
+            if (!TryGetSelectedMasteryRow(out var row) || row.QueueIndex is not > 0)
+                return;
+
+            var masteryId = row.MasteryId;
+            var ok = Mastery.ReorderTrainingQueueEntry(_playerId, row.QueueIndex.Value, direction, GetName(Player), GetPCPublicCDKey(Player), DateTime.UtcNow);
+
+            MasteryActionStatusText = ok ? "Queue reordered." : "Unable to reorder.";
+            LoadTargetMasteries();
+            ReselectMasteryRow(masteryId);
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs:984]: Fixed in 463b8db53: MoveQueueEntry re-resolves the current queue index by MasteryId+TargetTier at click time (same pattern as the abandon handler) and no-ops with a message if the entry no longer exists.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteriesViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+
+            return BackgroundJob.EnqueueDiscordWebhook(
+                webhookUrl,
+                characterName,
+                body,
+                15158332,
+                title,
+                createThread: true,
+                threadName: threadName);
+        }
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteriesViewModel.cs:817]: Fixed in 463b8db53: OnClickSubmitRequest now captures the Discord enqueue result in a try/catch, logs failures via LogGroup.Mastery, and surfaces a non-blocking warning instead of treating a failed enqueue as success.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteryReviewViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,1012 @@
...
+                    violations.Count > 0 ? overrideReason : string.Empty,
+                    useQuickSlot,
+                    isInstant,
+                    DateTime.UtcNow);
+
+                StatusMessageText = "Request approved.";
+                FeedbackText = string.Empty;
+                OverrideReasonText = string.Empty;
+                LoadQueue();
+            });
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteryReviewViewModel.cs:781]: Fixed in 463b8db53: custom-mastery catalog creation moved out of the VM modal callback into Mastery.ApproveRequest, gated by the same fresh Pending/InReview status re-check, so a stale reviewer window can no longer create a duplicate catalog row or overwrite a newer status.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteryReviewViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,1012 @@
...
+
+            _catalogPage = 0;
+            LoadCatalogManage();
+
+            var index = _catalogPageRows.FindIndex(m => m.Id == created.Id);
+            if (index < 0) return;
+
+            _catalogSelectedIndex = index;
+            CatalogManageToggles[index] = true;
+            LoadCatalogEditFields();
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteryReviewViewModel.cs]: Fixed in 463b8db53: "New Mastery" is now a local unsaved draft; CreateMastery is only called on Save after name/duplicate validation. Repeated clicks no longer create orphan placeholder rows.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+            bool useQuickSlot,
+            bool isInstant,
+            DateTime utcNow)
+        {
+            var request = DB.Get<MasteryRequest>(requestId);
+            if (request == null)
+                return false;
+
+            if (request.Status == MasteryRequestStatus.Approved || request.Status == MasteryRequestStatus.Denied)
+                return false;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs]: Fixed in 463b8db53: added MasteryRules.CanReviewRequest whitelisting Pending/InReview; both ApproveRequest and DenyRequest use it, so Cancelled requests can no longer be reviewed.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+                reviewFeedback,
+                request.Id,
+                utcNow);
+
+            request.Status = MasteryRequestStatus.Approved;
+            request.ReviewerName = actorName;
+            request.ReviewerCDKey = actorCDKey;
+            request.DateReviewed = utcNow;
+            request.ReviewFeedback = reviewFeedback ?? string.Empty;
+            request.OverrideReason = overrideReason ?? string.Empty;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs:312]: Fixed in 463b8db53: ApproveRequest re-runs ValidateRequest against current state immediately before mutation - blocking violations reject outright, and any warning now requires a non-empty override reason at the service layer, not just the VM.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+
+            DB.Set(profile);
+            DB.Set(request);
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs]: Fixed in 463b8db53: the request is persisted as Approved before the profile/training-queue write, so a partial failure can never double-enqueue or double-spend on retry (ordering constraint documented in code). Full multi-key transactions aren't available in this Redis persistence layer.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 9]
```diff
@@ -0,0 +1,641 @@
...
+                    return (MasteryTrainingSource.Retrain7, RetrainCredit7DurationDays);
+
+                if (profile.RetrainCredits14 > 0)
+                    return (MasteryTrainingSource.Retrain14, RetrainCredit14DurationDays);
+
+                // No credit actually available - fall through to normal resolution below.
+            }
+
+            if (useQuickSlot)
+                return (MasteryTrainingSource.QuickSlot, QuickSlotDurationDays);
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:83]: Fixed in 463b8db53: EnqueueTraining rejects useQuickSlot when QuickSlotsAvailable <= 0 before any mutation, and ApproveRequest aborts on that rejection. Zero-slot regression test added.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+
+        /// <summary>
+        /// The projected total level count once every currently queued/active training
+        /// entry completes (earned levels + one per queued entry, since each entry
+        /// advances its mastery by exactly one tier).
+        /// </summary>
+        public static int GetProjectedLevelTotal(PlayerMasteryProfile profile)
+        {
+            return GetEarnedLevelTotal(profile) + profile.TrainingQueue.Count;
+        }
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:260]: Fixed in 463b8db53: GetProjectedLevelTotal now projects the highest resulting tier per mastery (earned vs max queued TargetTier) instead of counting +1 per entry, with a prospective-request overload for the cap check. Regression test covers an overridden multi-tier jump.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+            }
+
+            var currentTier = profile.Masteries.TryGetValue(mastery.Id, out var level) ? level.Tier : 0;
+            if (targetTier != currentTier + 1)
+            {
+                violations.Add(new MasteryRuleViolation(
+                    MasteryRuleType.TierProgression,
+                    $"Requested tier {targetTier} is not the next tier for this mastery (current tier is {currentTier}).",
+                    false));
+            }
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:351]: Fixed in 463b8db53: added a blocking InvalidTier check for targetTier < 1 or > 5 in ValidateRequest, with tests for tiers 0 and 6.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 12]
```diff
@@ -0,0 +1,641 @@
...
+        /// Appends a "Reduce" audit entry.
+        /// </summary>
+        public static bool ReduceActiveTrainingTime(PlayerMasteryProfile profile, int days, MasteryActor actor, string reason, DateTime utcNow)
+        {
+            if (profile.TrainingQueue.Count == 0)
+                return false;
+
+            profile.TrainingQueue[0].ReductionDays += days;
+            AppendAudit(profile, utcNow, actor, "Reduce", reason);
+            return true;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:547]: Fixed in 463b8db53: ReduceActiveTrainingTime rejects non-positive values and clamps reductions to the entry's remaining duration so the finish date can never precede StartDate. Tests added for both.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 13]
```diff
@@ -0,0 +1,641 @@
...
+                {
+                    Name = seed.Name,
+                    Category = seed.Category,
+                    Description = seed.Description,
+                    Rarity = seed.Rarity,
+                    AssociatedSkill = seed.AssociatedSkill,
+                    IsActive = true,
+                    IsSeeded = true
+                })
+                .ToList();
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:804]: Fixed in 463b8db53: added an immutable SeedKey to the Mastery entity; seeding matches by SeedKey first (Name fallback for rows without one) and UpdateMastery preserves it, so renaming a seeded row no longer resurrects the original on restart. Renamed-seed regression test added.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+            bool useQuickSlot,
+            bool useRetrainCredit,
+            bool isInstant,
+            MasteryActor actor,
+            string reason,
+            string requestId,
+            DateTime utcNow)
+        {
+            if (useQuickSlot && profile.QuickSlotsAvailable <= 0)
+                return null;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs]: Fixed in ed3e2c5ff: the zero-Quick-Slot guard now runs only when the grant is not instant (`!isInstant && useQuickSlot && QuickSlotsAvailable <= 0`), so instant grants succeed regardless of a stale useQuickSlot flag. Added a regression test covering isInstant=true, useQuickSlot=true, zero slots: tier granted immediately, queue untouched, no slot consumed.

**Comment by zunath** [comment]
@coderabbitai full review

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 15]
```diff
@@ -0,0 +1,747 @@
...
+        /// <param name="utcNow">The current UTC time.</param>
+        /// <returns>True if the move was legal and applied; false otherwise.</returns>
+        public static bool ReorderQueueEntry(PlayerMasteryProfile profile, int index, int direction, MasteryActor actor, DateTime utcNow)
+        {
+            if (index <= 0 || index >= profile.TrainingQueue.Count)
+                return false;
+
+            var newIndex = index + direction;
+            if (newIndex <= 0 || newIndex >= profile.TrainingQueue.Count)
+                return false;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs:739]: Fixed in f7d520cbe2: ReorderQueueEntry rejects any direction other than -1/+1 before touching the queue - no self-swap or false Reorder audit entry on 0. Tests cover 0 and +/-2.

File: SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs [hunk 16]
```diff
@@ -0,0 +1,747 @@
...
+
+            if (useQuickSlot)
+                return (MasteryTrainingSource.QuickSlot, QuickSlotDurationDays);
+
+            return profile.LifetimeLevelsTrained switch
+            {
+                0 => (MasteryTrainingSource.Standard14, Standard14DurationDays),
+                1 => (MasteryTrainingSource.Standard21, Standard21DurationDays),
+                _ => (MasteryTrainingSource.Standard28, Standard28DurationDays)
+            };
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/MasteryService/MasteryRules.cs]: Fixed in f7d520cbe2: the duration bracket now derives from LifetimeLevelsTrained + TrainingQueue.Count, so back-to-back approvals get 14/21/28 rather than all landing in the first bracket, and a cancelled queued entry frees its slot naturally. The review window's duration preview shares this code path so it stays in sync. Regression tests cover sequential approvals and cancel-frees-slot.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 17]
```diff
@@ -0,0 +1,754 @@
...
+            var profile = GetOrCreateProfile(playerId);
+
+            if (profile.PendingCompletionNotices.Count == 0)
+                return new List<string>();
+
+            var notices = new List<string>(profile.PendingCompletionNotices);
+            profile.PendingCompletionNotices.Clear();
+            DB.Set(profile);
+
+            return notices;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs]: Fixed in f7d520cbe2: split into PeekPendingCompletionNotices (read-only) and AcknowledgeCompletionNotices (clear+persist); both callers now toast first and acknowledge after, so a UI exception can no longer lose notices.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 18]
```diff
@@ -0,0 +1,754 @@
...
+                actor,
+                reviewFeedback,
+                request.Id,
+                utcNow);
+
+            // EnqueueTraining rejects a Quick Slot request with none available before
+            // mutating anything - a stale or direct call must not be able to approve into
+            // a free discounted duration.
+            if (enqueued == null)
+                return false;
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs:305]: Fixed in f7d520cbe2: a shared MasteryRules.CanUseQuickSlot gate is checked in ApproveRequest before CreateMastery runs, so a zero-slot custom approval never persists an orphaned catalog row. Covered by a source regression test.

File: SWLOR.Game.Server/Service/Mastery.cs [hunk 19]
```diff
@@ -0,0 +1,754 @@
...
+
+            // A Custom (unlisted) request has no catalog row yet - build the same
+            // transient stand-in ValidateRequest/BuildEligibilityChecks use elsewhere so
+            // the rules re-check below applies identically to catalog and unlisted requests.
+            var existingMastery = string.IsNullOrWhiteSpace(request.MasteryId) ? null : GetMastery(request.MasteryId);
+            var checkMastery = existingMastery ?? new Entity.Mastery
+            {
+                Name = request.CustomName,
+                Rarity = MasteryRarityType.Standard
+            };
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Service/Mastery.cs:235]: Fixed in f7d520cbe2: ApproveRequest now rejects outright when a non-custom request's catalog row no longer resolves, instead of treating it as a custom stand-in. A retired-but-present (IsActive=false) row remains approvable by design - staff already saw and accepted the request; the code documents this.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteriesViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,925 @@
...
+                SendMessageToPC(Player, ColorToken.Green("Mastery request submitted! Staff will review it soon."));
+                SendMessageToPC(Player, ColorToken.Orange("(Staff Discord notification could not be sent - staff will still see this in /masteryreview.)"));
+            }
+            else
+            {
+                SendMessageToPC(Player, ColorToken.Green("Mastery request submitted! Staff will review it soon."));
+            }
+
+            ShowMyRequests();
+        };
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/MasteriesViewModel.cs:767]: Fixed in f7d520cbe2: added an in-flight guard on OnClickSubmitRequest (submit disabled until the handler completes), plus service-level defense-in-depth - Mastery.SubmitRequest rejects an identical Pending/InReview request for the same player/mastery/tier via MasteryRules.FindDuplicatePendingRequest. Dedup is unit-tested.

File: SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs [hunk 21]
```diff
@@ -378,5 +594,415 @@ public Action OnClickSaveChanges() => () =>
...
+                return;
+            }
+
+            var masteryId = row.MasteryId;
+            Mastery.ReduceTrainingTime(_playerId, days, GetName(Player), GetPCPublicCDKey(Player), MasteryActionReason, DateTime.UtcNow);
+
+            MasteryActionStatusText = $"Reduced active training by {days} day(s).";
+            LoadTargetMasteries();
+            ReselectMasteryRow(masteryId);
+        };
```
[the remark this answers is not in the document]
Comment (zunath) [reply SWLOR.Game.Server/Feature/GuiDefinition/ViewModel/DMPlayerExamineViewModel.cs:910]: Fixed in f7d520cbe2: the reduce handler re-fetches the profile at click time and verifies the clicked row (MasteryId+TargetTier) is still the active entry before reducing; otherwise it shows a "training state changed" message and does nothing - same re-resolution pattern as the abandon/reorder handlers.
